src/lib_c/x86_64-windows-msvc/c/winapi.cr

```diff
@@ -0,0 +1,241 @@
+lib LibC
+  STD_INPUT_HANDLE  = 0xFFFFFFF6_u32
+  STD_OUTPUT_HANDLE = 0xFFFFFFF5_u32
+  STD_ERROR_HANDLE  = 0xFFFFFFF4_u32
+
+  FILE_TYPE_UNKNOWN = 0x0000
+  FILE_TYPE_DISK    = 0x0001
+  FILE_TYPE_CHAR    = 0x0002
+  FILE_TYPE_PIPE    = 0x0003
+  FILE_TYPE_REMOTE  = 0x8000
+
+  GENERIC_EXECUTE = 0x20000000
+  GENERIC_WRITE   = 0x40000000
+  GENERIC_READ    = 0x80000000
+
+  CREATE_NEW        = 1
+  CREATE_ALWAYS     = 2
+  OPEN_EXISTING     = 3
+  OPEN_ALWAYS       = 4
+  TRUNCATE_EXISTING = 5
+
+  FILE_FLAG_OVERLAPPED = 0x40000000
+
+  # source https://msdn.microsoft.com/en-us/library/windows/desktop/aa383751(v=vs.85).aspx
+  alias Long = Int32
+  alias Word = UInt16
+  alias WChar = UInt16
+  alias DWord = UInt32
+  alias Handle = Void*
+  alias SizeT = UInt64
+  alias BOOL = Int32 # FIXME maybe it need to be removed because it can be confused with Bool
+
+  INVALID_HANDLE_VALUE = Pointer(Void).new((-1).to_u64)
+  INFINITY             = 0xFFFFFFFF_u32
+
+  fun _DuplicateHandle = DuplicateHandle(source_process : Handle, source : Handle, target_process : Handle, target : Handle*, desired_access : DWord, inherit_handle : Bool, options : DWord) : Bool
+
+  # member names are not original otherwise they would be really confusing
+  struct OVERLAPPED
+    status : SizeT
+    bytes_transfered : SizeT
+    offset : UInt64
+    event : Handle
+  end
+
+  struct SECURITY_ATTRIBUTES
+    nLength : DWord
+    lpSecurityDescriptor : Void*
+    bInheritHandle : BOOL
+  end
+
+  fun _GetStdHandle = GetStdHandle(std_handle : DWord) : Handle
```

Comment:
I'm confused as to why we're *adding* underscores to symbol names which appear not to be underscored in the first place.

Reply:
I think we could relax this rule, I have no problem with that, specially for lib functions. For class methods there's also no problem. The only problem is top level methods:

```crystal
# This is valid in Ruby
def Foo(x)
end

# But this is always a generic instantiation in Crystal
Foo(String)
```

We still have time to relax this rule everywhere and change generics to use `<...>`. There's a bit of ambiguity if the parser finds `Foo<` and it doesn't know if it's a generic instantiation or a comparison, but we could make it so that `Foo<` (no space between type name and `<`) is always a generic instantiation, and problem solved.

For now, we can relax the rule for lib funs and relax the rule for method calls...

Reply:
I'd prefer if if we only relaxed the rule for lib funs myself, I don't think there's a usecase for this outside external symbols, and I don't want to encourage confusing method casing for typical class maethods.

Reply:
I'd prefer to change the compiler, though only for `fun`s for now. It will also make it easier to automatically generate bindings, not having to translate names to other names (which right now sometimes result in conflicts). It's also simpler to understand if you know the C functions.

Reply:
Personally I'd prefer `snake_case` casing, since it feels more Crystal-ish, yet it might involve problems with fun name de/normalization.

Reply:
If we allow any case for `fun`s, shouldn't we allow any case for `struct`s inside `Lib`?
And that will imply allowing that change in the type restrictions for the `fun`'s arguments/return types.
If we aim to simplify binding generations I think that will be needed as well.

I thought there could be a problem with that if in the future we allow bindings to C++, since there will be a need for generic class inside the `Lib`. But it turns out that fun calls are written as `LibA.Method(arg)` and types would be `LibA::Type(arg)` so there is no problem here. As long as we keep not permitting an `include LibA`. Allowing `include LibA` would leave us flat in knowing if `Foo(1).bar` is a fun and method call or a class method.

Reply:
@bcardiff I was thinking exactly the same thing yesterday. This should be applicable to structs, and maybe even aliases like `size_t`. But it's true that for restrictions it's a bit tricky, though we could maybe parse restrictions differently inside lib declarations. This will not only make it easier to automatically generate bindings, but also simpler to copy C headers without having to translate `size_t` to `SizeT`, etc.

Since all these types have to be prefixed with a lib, like `LibC::SizeT`, I think we can introduce the rule for `LibC::size_t` parsing fine. And for restrictions inside lib declarations we can allow `fun foo(x : size_t)`.

This also removes the confusion between Crystal's `Int` and libc's `int`, which we could just have as `int`.

Yesterday I also remembered we have `@[External]` so that crystal structs can be used in C libraries, so I thought maybe these should also have similar names, but since it's on Crystal's side I think it's fine if we require these to start with uppercase.

Reply:
I still think that this is a lot of unnecessary complexity just to allow us to use names from C unchanged. Is it really worth it?

Reply:
I think it's almost trivial to implement and will simplify writing bindings. So why not?